SAPT(DFT) MO disk I/O optimization & Exchange-Dispersion scaling scheme update

yxie326 opened the pull request:
## Description
Optimizes the integral transformation step of SAPT(DFT), in which the transformed MO needs to be written to the disk. The `STORE` scheme in `src/psi4/lib3index/dfhelper.cc`  has problem in writing blocks of integrals efficiently, and this PR changes it to `direct_iaQ` to optimize the writing process. Also fixed a few memory related bugs and modified a few timer tags.
The deafult scaling scheme of SAPT(DFT) exchange-dispersion energy is now changed from `DISP` to `FIXED`. It scales the uncoupled Exch-Disp2 by 0.769848. Deeper details are discussed in the paper [Y. Xie, D. G. A. Smith, and C. D. Sherrill, J. Chem. Phys. **157**, 024801 (2022)].
The SAPT(DFT) procedure is also optimized for the case that `SAPT_DFT_FUNCTIONAL = HF`, i.e. running SAPT0 with the SAPT(DFT) driver, to avoid redundant SCF calculations.

## Todos

- [ ] Optimization of disk I/O in SAPT(DFT) integral transformation
- [ ] Fixed minor memory related bugs and modified timer tags
- [ ] Switched default exch-disp scaling scheme to FIXED (default factor 0.769848)
- [ ] Optimization of SAPT(HF) procedure

## Questions
- [ ] Question1

## Checklist
- [ ] Tests added for any new features
- [ ] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [ ] Ready for merge

JonathonMisiewicz commented:
This PR includes your work-in-progress implementation of automatic GRAC shifts, which is not ready for review. Please revise the PR to only include the code _that is actually supposed to be part of the PR_.

yxie326 commented:
> This PR includes your work-in-progress implementation of automatic GRAC shifts, which is not ready for review. Please revise the PR to only include the code _that is actually supposed to be part of the PR_.

Done. I've completely removed the automatic GRAC shift computations, because this could be non-trivial due to uncertain SCF converging issues, and I and Dr. Sherrill have agreed to postpone this feature until we find a better solution.

JonathonMisiewicz commented on psi4/src/psi4/lib3index/dfhelper.h:
```diff
@@ -304,6 +304,9 @@ class PSI_API DFHelper {
     void clear_transformations(); 
 
+    /// clear in-core AO integral
+    void clear_AO();
```
Why define this function if you never call it?

yxie326 replied:
This was originally in `psi4/src/psi4/libsapt_solver/fdds_disp.cc`, it was created so that the program can clear the AO buffer outside of the dfhelper module. It was once needed during one of the attempts to fix the MO disk I/O issue, but it's not needed anymore in the final version (clearing AO buffer is done within `transform()` of the dfhelper module instead). Should I still keep this function here?

JonathonMisiewicz replied:
No. If you don't use a function and don't plan to use the function, don't keep it.

yxie326 commented on psi4/driver/procrouting/sapt/sapt_proc.py:
```diff
@@ -70,6 +70,7 @@ def run_sapt_dft(name, **kwargs):
     mon_b_shift = core.get_option("SAPT", "SAPT_DFT_GRAC_SHIFT_B")
     do_delta_hf = core.get_option("SAPT", "SAPT_DFT_DO_DHF")
     sapt_dft_functional = core.get_option("SAPT", "SAPT_DFT_FUNCTIONAL")
+    do_dft = sapt_dft_functional != "HF"
```
This is to optimize running SAPT0 with the SAPT(DFT) module by setting DFT functional to HF. In SAPT(DFT) the program does a delta-HF calculation based on HF orbitals and then followed by DFT monomer caluclations; for SAPT0 this will be redundant so I added some logics to avoid this.

JonathonMisiewicz replied:
Could we determine if this is _actually_ a DFT computation or not at the Python layer?

yxie326 replied:
I'm not sure what you mean... Do you mean determining if it's a HF or DFT calculation? I think HF and DFT shares the same driver function `run_scf` in Psi4

JonathonMisiewicz replied:
No. I mean that if the user requests to do SAPT(DFT) using Hartree-Fock, the code should direct them to use the normal SAPT(0) code, if that's a more efficient way to do the same computation.

loriab replied:
Doing a SAPT0 is going to get more complicated before it gets simpler. Right now there are three ways -- Ed, Rob, Daniel, in short. And with some of the more-accurate non-S^2-approx terms that the Patkowski group has implemented, there's good coding reasons that one would run a SAPT0 through the SAPT(DFT) aka "Daniel" route. It makes for unlovely input files to get the highest-level SAPT calc possible from psi. I think Yi's special-casing the HF fctl may be part of figuring out a more rational future for SAPT code routing.

In general, I agree with Jonathon's advice on steering users toward the most efficient route. But here I don't think we want to explain the present or commit to the future. And if users have already strayed this far from `energy("sapt0")`, I think let them continue.

JonathonMisiewicz replied:
That's fine with me, but can you or Yi write an issue that can say what the three ways are, what this means for short-term SAPT development, and (eventually) coordinate plans for improving the situation? This problem is going to come up frequently enough that it deserves a dedicated reference.

PhillCli replied:
One thing to consider is, AFAIK, running SAPT0 through the SAPT(DFT) path computes coupled dispersion (orbital relaxation effects) at HF level which none of the alternative routes have implemented.

yxie326 replied:
@PhillCli Exactly. One of my next plan is to make coupled dispersion an option so that the code would be more useful with SAPT0, since SAPT0 does not depend on coupled dispersion as heavily as SAPT(DFT) does. As Lori said, one main reason that users would choose to use the "Daniel" code for SAPT0 is that it seems to be the only one that has support to non-S^2-approximated induction & exchange-induction terms.

JonathonMisiewicz replied:
For future reference, the requested notes exist but are not public access. If you need this, contact Lori directly.

yxie326 commented on psi4/src/read_options.cc:
```diff
@@ -1040,9 +1040,9 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
         ``NONE`` Use unscaled ``Exch-Disp2,u`` .
         ``FIXED`` Use a fixed factor |sapt__sapt_dft_exch_disp_fixed_scale| to scale ``Exch-Disp2,u`` .
         ``DISP`` Use the ratio of ``Disp2,r`` and ``Disp2,u`` to scale ``Exch-Disp2,u`` . -*/
-        options.add_str("SAPT_DFT_EXCH_DISP_SCALE_SCHEME", "DISP", "NONE FIXED DISP");
```
Thanks for pointing out; I'll edit the contents of PR.

yxie326 commented on psi4/src/psi4/libsapt_solver/fdds_disp.cc:
```diff
@@ -213,25 +217,26 @@ FDDS_Dispersion::FDDS_Dispersion(std::shared_ptr<BasisSet> primary, std::shared_
     }
 
     dfh_->clear_spaces();
+    // dfh_->clear_AO();
```
This is what I mentioned in `dfhelper.h`. Maybe removing everything related to this `clear_AO()` as if it never existed would be less confusing to other contributors...

JonathonMisiewicz commented:
Ping me for review again when you have this passing tests.

JonathonMisiewicz commented:
While I appreciate that tests are now passing, I still request changes.
* Comments that refer to methods that no longer exist need to be removed.
* The comment on dfhelper's release_AO is vague. What does it mean to "release AO"? Are you clearing memory used to store some AO-basis quantity?
* Naming a method "release_AO" that does not actually release AO is confusing. You need to change this.
* The title of this PR is misleading. This PR is not just optimizing I/O. This PR is _changing the values that Psi gives for SAPT(DFT)_ due to the options edits. Are these edits supposed to be part of the PR or not? That's going to determine how the rest of this review process goes.

yxie326 commented:
> Comments that refer to methods that no longer exist need to be removed.

I'm not quite sure what do they refer to...

JonathonMisiewicz commented on psi4/src/psi4/lib3index/dfhelper.cc:
```diff
@@ -1951,8 +1951,9 @@ void DFHelper::transform() {
                     // get transformation info
-                    std::string left = std::get<0>(transf_[order_[count + k]]);
-                    std::string right = std::get<1>(transf_[order_[count + k]]);
+                    std::string transf_name = order_[count + k];
+                    std::string left = std::get<0>(transf_[transf_name]);
+                    std::string right = std::get<1>(transf_[transf_name]);
```
```suggestion
                    auto transf_name = order_[count + k];
                    auto left = std::get<0>(transf_[transf_name]);
                    auto right = std::get<1>(transf_[transf_name]);
```

JonathonMisiewicz commented on psi4/src/psi4/lib3index/dfhelper.cc:
```diff
@@ -2015,7 +2016,13 @@ void DFHelper::transform() {
 
     // transformations complete, time for metric contractions
```
What is the transform function trying to accomplish? The function isn't properly documented, and I can't review the code without that.

yxie326 replied:
This transform function converts 3-index integrals with AO functions (pq|Q) to those with MO functions. The details of how they should be transformed are configured via other functions like `add_space()` and `add_transformation()`, and `transform()` performs the actual transformation with the configurations defined above.

JonathonMisiewicz reviewed (dismissed):
I want @CDSherrill to confirm that these changes to the SAPT parameters are meant to come in now.

The fact that you can change this parameter _without breaking tests_ indicate that our SAPT(DFT) tests are not good enough.

loriab commented on psi4/src/psi4/lib3index/dfhelper.h:
```diff
@@ -105,6 +105,12 @@ class PSI_API DFHelper {
+    /// intermediates for the metric contraction. Used in methods 
+    /// that AO integrals are not needed in the contraction step 
+    /// like SAPT(DFT). 
+    void switch_release_AO() { release_AO_ = true; }
```
I don't know if `release_AO` is best set up. Default is false, and this switch always changes it to true. But the fn name is just "switch" which could imply `release_AO_ = ! release_AO_`. Perhaps a regular setter would be clearer.

yxie326 replied:
Thanks, changed it to a setter. It might be an issue in English since "switch" is what pops up in my head when trying to do this :)

loriab reviewed (commented):
partial review.

yxie326 commented:
> I want @CDSherrill to confirm that these changes to the SAPT parameters are meant to come in now.
> 
> The fact that you can change this parameter _without breaking tests_ indicate that our SAPT(DFT) tests are not good enough.

I think we have manually set `set SAPT_DFT_EXCH_DISP_SCALE_SCHEME disp` in the previous SAPT(DFT) tests, and that's why changing the default value of `SAPT_DFT_EXCH_DISP_SCALE_SCHEME` from `disp` to `fixed` won't break the tests. 
Should we update these tests with `SAPT_DFT_EXCH_DISP_SCALE_SCHEME` set to `fixed`?

JonathonMisiewicz commented:
Leave those tests as they are, and create new tests for the fixed option.

I also insist that you include a warning about those keyword changes in the documentation for SAPT(DFT), and in the output or logging file whenever SAPT(DFT) runs. This could cause users to get different numbers with the exact same input file.

yxie326 commented:
> 

Created new test cases in the same input file, i.e. doing non-hybrid/hybrid+DISP/hybrid+FIXED in the same test file. I think this is the way to go since it is what we did when adding in the hybrid feature.

Warning messages added in documentation and SAPT(DFT) output.

yxie326 commented on psi4/src/psi4/lib3index/dfhelper.h:
```diff
@@ -105,6 +105,9 @@ class PSI_API DFHelper {
     void set_AO_core(bool core) { AO_core_ = core; }
     bool get_AO_core() { return AO_core_; }
 
+    /// Switch flag for releasing AO (for SAPT(DFT))
```
The function `transform()` consists of 2 steps:
1. Transforms 3-index integrals with AO basis (pq|Q) to those with MO basis such as (ia|Q)
2. Contract it with a power of metric (since it will be useful in some density fitting algorithms). Most common choice of power would be -1/2 and this contraction would yield (ia|Q)(Q|Q')^(-1/2). 

In SAPT(DFT) hybrid dispersion, when in-core AO is turned on, the AO integral (pq|Q) is saved in the memory after step 1, but it is not needed in step 2, and releasing it allows much better blockings in step 2. (In fact what actually happened is that the memory limit was exceeded when trying to run large calculations before allowing the release of (pq|Q) after step 1)

However, it seems that when we release it, it will cause some other modules to fail, in particular Rob's FISAPT program for computing exchange-dispersion energy. This might be because in FISAPT (pq|Q) are still needed in step 2, but I have not figured out why this is the case. To avoid further complexity, I decided to release (pq|Q) only in SAPT(DFT) hybrid dispersion by using this switch flag.

yxie326 commented:
> core developers

Ready for review now, thanks.

CDSherrill commented:
I've discussed this PR with Yi.  This is just to confirm that we want to change the default exch-disp scaling scheme, as the new one appears to be more reliable.  Apparently the previous test case didn't break because the exch-disp scheme was manually selected in the test case, rather than utilizing the (updated) default.

JonathonMisiewicz commented on psi4/src/psi4/libmints/matrix.cc:
```diff
@@ -1459,7 +1459,8 @@ void Matrix::axpy(double a, SharedMatrix X) {
     for (int h = 0; h < nirrep_; h++) {
         size_t size = colspi_[h ^ symmetry()] * (size_t)rowspi_[h];
-        if (size != (X->rowspi()[h] * X->colspi()[h ^ X->symmetry()])) {
+        size_t size_X = X->colspi()[h ^ X->symmetry()] * (size_t)X->rowspi()[h];
```
Why is this change necessary? Was there some issue with type casting?

JonathonMisiewicz commented on doc/sphinxman/source/sapt.rst:
```diff
@@ -532,9 +532,13 @@ estimated by scaling from the uncoupled value either by a fitted fixed
+.. warning:: The default option of |sapt__sapt_dft_exch_disp_scale_scheme|
+             has been changed from ``DISP`` to ``FIXED``. This might cause
+             an older version of |PSIfour| to obatin a different value of
+             exchange-dispersion energy from the latest version.
```
There also needs to be a warning about the change to "SAPT_DFT_EXCH_DISP_FIXED_SCALE", both here in the documentation and in the output file.

JonathonMisiewicz reviewed (changes requested):
We have no test coverage on the `DFT_DO_DHF` keyword, and this PR changes its behavior.

I'm going to request a _separate_ PR that adds this test coverage. That PR must be merged in first, so that we can be sure this PR is not changing the results from that keyword.

yxie326 commented:
> We have no test coverage on the `DFT_DO_DHF` keyword, and this PR changes its behavior.
> 
> I'm going to request a _separate_ PR that adds this test coverage. That PR must be merged in first, so that we can be sure this PR is not changing the results from that keyword.

Is a few new test cases everything that new PR should have, or should I cherry-pick the changes related to the `DFT_DO_DHF` keyword in this PR and apply to that one?

loriab reviewed (commented):
Thanks for the updates, Yi. Here's a few suggestions on notating the defaults changes.

loriab commented:
Hi Yi,

Thanks for your work and the comments. I've been discussing some with Jonathon, and I think if you could do the below, that will clarify all the scaling factor and testing aspects of the PR, so we can move on to the I/O optimization and routing logic parts. Please let me know of any concerns.

* accept the various GH suggestions clarifying the scaling scheme changes (only if you agree with them, of course). Feel free to make the changes locally, rather than through the GH interface, if that makes the git operations easier.
* edit sapt-dft1 test to be something like the below. This adds an extra section without the deltaHF correction. The analogous value in test sapt-dft2 is `DHF = -1.42620815`. This passes on master for me, so having the dHF=False pass on your PR will help verify the separate routing logic you added. With this in place, there won't be a need to separate out tests into another PR.
```
#! SAPT(DFT) aug-cc-pVDZ interaction energy between Ne and Ar atoms.

DHF = -0.01189736  #TEST
Eref_nh = {"SAPT ELST ENERGY":  -0.10190449, #TEST
           "SAPT EXCH ENERGY":   0.36545706, #TEST
           "SAPT IND ENERGY":   -0.00840483, #TEST
           "SAPT DISP ENERGY":  -0.24398704, #TEST
           "CURRENT ENERGY":     0.01122234} #TEST

Eref_h_disp = {"SAPT ELST ENERGY":  -0.10197193, #TEST
               "SAPT EXCH ENERGY":   0.36569812, #TEST
               "SAPT IND ENERGY":   -0.00840370, #TEST
               "SAPT DISP ENERGY":  -0.26658499, #TEST
               "CURRENT ENERGY":    -0.01126250} #TEST

Eref_h_fixed = {"SAPT ELST ENERGY":  -0.10197193, #TEST
                "SAPT EXCH ENERGY":   0.36569812, #TEST
                "SAPT IND ENERGY":   -0.00840370, #TEST
                "SAPT DISP ENERGY":  -0.26605283, #TEST
                "CURRENT ENERGY":    -0.01073034} #TEST

molecule dimer {
  Ne
  --
  Ar 1 6.5
  units bohr
}

set {
    basis         aug-cc-pvdz
    scf_type      df
    sapt_dft_grac_shift_a 0.203293
    sapt_dft_grac_shift_b 0.138264
}

# No hybrid kernel & no exch-disp scaling & no deltaHF
set SAPT_DFT_DO_DHF False
set SAPT_DFT_DO_HYBRID False
set SAPT_DFT_EXCH_DISP_SCALE_SCHEME none
energy('sapt(dft)', molecule=dimer)
for k, v in Eref_nh.items():                           #TEST
    if k in ["SAPT IND ENERGY", "CURRENT ENERGY"]:
        ref = (v - DHF) / 1000.0
    else:
        ref = v / 1000.0
    compare_values(ref, psi4.variable(k), 6, "!hyb, xd=none, !dHF: " + k) #TEST

# No hybrid kernel & no exch-disp scaling
set SAPT_DFT_DO_DHF True
set SAPT_DFT_DO_HYBRID False
set SAPT_DFT_EXCH_DISP_SCALE_SCHEME none
energy('sapt(dft)', molecule=dimer)
for k, v in Eref_nh.items():                           #TEST
    compare_values(v / 1000.0, psi4.variable(k), 6, "!hyb, xd=none, dHF: " + k) #TEST

# Hybrid kernel & exch-disp scaling (DISP)
set SAPT_DFT_DO_HYBRID True
set SAPT_DFT_EXCH_DISP_SCALE_SCHEME disp
energy('sapt(dft)', molecule=dimer)
for k, v in Eref_h_disp.items():                       #TEST
    compare_values(v / 1000.0, psi4.variable(k), 6, "hyb, xd=disp, dHF: " + k) #TEST

# Hybrid kernel & exch-disp scaling (FIXED)
set SAPT_DFT_DO_HYBRID True
set SAPT_DFT_EXCH_DISP_SCALE_SCHEME fixed 
set SAPT_DFT_EXCH_DISP_FIXED_SCALE 0.770
energy('sapt(dft)', molecule=dimer)
for k, v in Eref_h_fixed.items():                      #TEST
    compare_values(v / 1000.0, psi4.variable(k), 6, "hyb, xd=fixed, dHF: " + k) #TEST

```
* I think it should be clear from the output file what exch-disp scheme/scale is applied. So perhaps add a couple lines to the printout like below. This should also satisfy Jonathon's request that the output file show the change in scaling defaults.

```
  ==> E20 Dispersion (MP2) <==


    Disp20 (MP2)                     -0.37881730 [mEh]
    Exch-Disp20,u                     0.02037338 [mEh]
    Scaling Scheme:                   Disp
    Scaling Factor:                   0.707

   SAPT(DFT) Results
  ---------------------------------------------------------------------------------------------------------
    Electrostatics                   -0.10197192 [mEh]     -0.06398835 [kcal/mol]     -0.26772724 [kJ/mol]
      Elst1,r                        -0.10197192 [mEh]     -0.06398835 [kcal/mol]     -0.26772724 [kJ/mol]
```

yxie326 commented on psi4/src/read_options.cc:
```diff
@@ -1055,9 +1055,11 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
-        options.add_double("SAPT_DFT_EXCH_DISP_FIXED_SCALE", 0.686);
+        options.add_str("SAPT_DFT_EXCH_DISP_SCALE_SCHEME", "FIXED", "NONE FIXED DISP");
+        /*- Exch-disp scaling factor for FIXED scheme for |sapt__sapt_dft_exch_disp_scale_scheme|.
+        Default value of 0.770 suggested in Y. Xie, D. G. A. Smith and C. D. Sherrill, 2022 (submitted).
```
The value in paper is 0.770; It is rounded from 0.769848.

hokru replied:
Then the default value below should be changed to 0.770.

yxie326 commented:
> Hi Yi,
> 
> Thanks for your work and the comments. I've been discussing some with Jonathon, and I think if you could do the below, that will clarify all the scaling factor and testing aspects of the PR, so we can move on to the I/O optimization and routing logic parts. Please let me know of any concerns.
> 
> * accept the various GH suggestions clarifying the scaling scheme changes (only if you agree with them, of course). Feel free to make the changes locally, rather than through the GH interface, if that makes the git operations easier.
> * edit sapt-dft1 test to be something like the below. This adds an extra section without the deltaHF correction. The analogous value in test sapt-dft2 is `DHF = -1.42620815`. This passes on master for me, so having the dHF=False pass on your PR will help verify the separate routing logic you added. With this in place, there won't be a need to separate out tests into another PR.
> 
> ```
> #! SAPT(DFT) aug-cc-pVDZ interaction energy between Ne and Ar atoms.
> 
> DHF = -0.01189736  #TEST
> Eref_nh = {"SAPT ELST ENERGY":  -0.10190449, #TEST
>            "SAPT EXCH ENERGY":   0.36545706, #TEST
>            "SAPT IND ENERGY":   -0.00840483, #TEST
>            "SAPT DISP ENERGY":  -0.24398704, #TEST
>            "CURRENT ENERGY":     0.01122234} #TEST
> 
> Eref_h_disp = {"SAPT ELST ENERGY":  -0.10197193, #TEST
>                "SAPT EXCH ENERGY":   0.36569812, #TEST
>                "SAPT IND ENERGY":   -0.00840370, #TEST
>                "SAPT DISP ENERGY":  -0.26658499, #TEST
>                "CURRENT ENERGY":    -0.01126250} #TEST
> 
> Eref_h_fixed = {"SAPT ELST ENERGY":  -0.10197193, #TEST
>                 "SAPT EXCH ENERGY":   0.36569812, #TEST
>                 "SAPT IND ENERGY":   -0.00840370, #TEST
>                 "SAPT DISP ENERGY":  -0.26605283, #TEST
>                 "CURRENT ENERGY":    -0.01073034} #TEST
> 
> molecule dimer {
>   Ne
>   --
>   Ar 1 6.5
>   units bohr
> }
> 
> set {
>     basis         aug-cc-pvdz
>     scf_type      df
>     sapt_dft_grac_shift_a 0.203293
>     sapt_dft_grac_shift_b 0.138264
> }
> 
> # No hybrid kernel & no exch-disp scaling & no deltaHF
> set SAPT_DFT_DO_DHF False
> set SAPT_DFT_DO_HYBRID False
> set SAPT_DFT_EXCH_DISP_SCALE_SCHEME none
> energy('sapt(dft)', molecule=dimer)
> for k, v in Eref_nh.items():                           #TEST
>     if k in ["SAPT IND ENERGY", "CURRENT ENERGY"]:
>         ref = (v - DHF) / 1000.0
>     else:
>         ref = v / 1000.0
>     compare_values(ref, psi4.variable(k), 6, "!hyb, xd=none, !dHF: " + k) #TEST
> 
> # No hybrid kernel & no exch-disp scaling
> set SAPT_DFT_DO_DHF True
> set SAPT_DFT_DO_HYBRID False
> set SAPT_DFT_EXCH_DISP_SCALE_SCHEME none
> energy('sapt(dft)', molecule=dimer)
> for k, v in Eref_nh.items():                           #TEST
>     compare_values(v / 1000.0, psi4.variable(k), 6, "!hyb, xd=none, dHF: " + k) #TEST
> 
> # Hybrid kernel & exch-disp scaling (DISP)
> set SAPT_DFT_DO_HYBRID True
> set SAPT_DFT_EXCH_DISP_SCALE_SCHEME disp
> energy('sapt(dft)', molecule=dimer)
> for k, v in Eref_h_disp.items():                       #TEST
>     compare_values(v / 1000.0, psi4.variable(k), 6, "hyb, xd=disp, dHF: " + k) #TEST
> 
> # Hybrid kernel & exch-disp scaling (FIXED)
> set SAPT_DFT_DO_HYBRID True
> set SAPT_DFT_EXCH_DISP_SCALE_SCHEME fixed 
> set SAPT_DFT_EXCH_DISP_FIXED_SCALE 0.770
> energy('sapt(dft)', molecule=dimer)
> for k, v in Eref_h_fixed.items():                      #TEST
>     compare_values(v / 1000.0, psi4.variable(k), 6, "hyb, xd=fixed, dHF: " + k) #TEST
> ```
> 
> * I think it should be clear from the output file what exch-disp scheme/scale is applied. So perhaps add a couple lines to the printout like below. This should also satisfy Jonathon's request that the output file show the change in scaling defaults.
> 
> ```
>   ==> E20 Dispersion (MP2) <==
> 
> 
>     Disp20 (MP2)                     -0.37881730 [mEh]
>     Exch-Disp20,u                     0.02037338 [mEh]
>     Scaling Scheme:                   Disp
>     Scaling Factor:                   0.707
> 
>    SAPT(DFT) Results
>   ---------------------------------------------------------------------------------------------------------
>     Electrostatics                   -0.10197192 [mEh]     -0.06398835 [kcal/mol]     -0.26772724 [kJ/mol]
>       Elst1,r                        -0.10197192 [mEh]     -0.06398835 [kcal/mol]     -0.26772724 [kJ/mol]
> ```

Added the !dHF cases in `sapt-dft1` and `sapt-dft2`. For `sapt-dft-api`, the dHF feature is intrinsically unsupported, because the dHF calculation is not fully contained in the sapt_dft() call; a part of it is in the run_sapt_dft() call. 

Added the scaling scheme/factor output lines in the output file.

loriab reviewed (commented):
Thanks! here's a couple more GH suggestions on the exch-disp.

yxie326 commented on psi4/driver/procrouting/sapt/sapt_proc.py:
```diff
@@ -432,7 +460,23 @@ def sapt_dft(dimer_wfn, wfn_A, wfn_B, sapt_jk=None, sapt_jk_B=None, data=None, p
+    elif exch_disp_scheme == "DISP":
+        exch_disp_scale = data["Disp20"] / data["Disp20,u"]
+        data["Exch-Disp20,r"] = exch_disp_scale * data["Exch-Disp20,u"]
+    core.print_out(print_sapt_var("Exch-Disp20,r", data["Exch-Disp20,r"], short=True) + "\n")
```
> This will match the printout in the sapt dft summary below, right? The "Est." should help with matching quantities and emphasize that it's not explicitly computed. Thanks for adding the options status.

Yes. I'll manually modify this since I'd like the program not to print out this line if exch_disp_scheme is NONE.

yxie326 commented on psi4/src/read_options.cc:
```diff
@@ -1052,12 +1052,15 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
+        /*- Exch-disp scaling factor for FIXED scheme for |sapt__sapt_dft_exch_disp_scale_scheme|.
+        Default value of 0.770 suggested in Y. Xie, D. G. A. Smith and C. D. Sherrill, 2022 (submitted).
+        Previous to Nov 2022, default value was 0.686 suggested by Hesselmann and Korona, J. Chem. Phys. 141, 094107 (2014). !expert -*/
+        options.add_double("SAPT_DFT_EXCH_DISP_FIXED_SCALE", 0.769848);
```
Okay. Probably this is the way to go because any accuracies after 3rd decimal place is meaningless because of the approximate nature of the scaling scheme. I think the reason I chose to put 3 decimal places in the paper is to follow what Hesselmann and Korona did in [Hesselmann and Korona, J. Chem. Phys. 141, 094107 (2014).].

loriab commented on psi4/src/psi4/lib3index/dfhelper.h:
```diff
@@ -335,6 +341,7 @@ class PSI_API DFHelper {
     bool symm_compute_;
     bool AO_core_ = true;
     bool MO_core_ = false;
+    bool release_AO_ = false;
```
```suggestion
    bool release_core_AO_before_metric_ = false;
```

loriab commented on psi4/src/psi4/libsapt_solver/fdds_disp.cc:
```diff
@@ -184,17 +186,18 @@ FDDS_Dispersion::FDDS_Dispersion(std::shared_ptr<BasisSet> primary, std::shared_
     }
 
     // transform
+    dfh_->set_release_AO(true);
```
```suggestion
    dfh_->set_release_core_AO_before_metric(true);
```

loriab commented on psi4/src/psi4/libsapt_solver/fdds_disp.cc:
```diff
@@ -209,29 +212,30 @@ FDDS_Dispersion::FDDS_Dispersion(std::shared_ptr<BasisSet> primary, std::shared_
         dfh_->add_transformation("bbR", "b", "b", "pqQ");
         dfh_->add_transformation("bsR", "b", "s", "pqQ");
         dfh_->add_transformation("ssR", "s", "s", "pqQ");
+        dfh_->set_release_AO(true);
```
```suggestion
        dfh_->set_release_core_AO_before_metric(true);
```

loriab commented on psi4/src/psi4/lib3index/dfhelper.cc:
```diff
@@ -2004,7 +2005,13 @@ void DFHelper::transform() {
+    timer_on("DFH: Metric Contractions");
+
+    // release in-core AO 
+    if (AO_core_ && release_AO_)  {
```
```suggestion
    if (AO_core_ && release_core_AO_before_metric_)  {
```

loriab commented on psi4/src/psi4/lib3index/dfhelper.cc:
```diff
@@ -2036,8 +2043,9 @@ void DFHelper::transform() {
 
             } else {
                 // total size allowed, in doubles
+                size_t rem_mem = memory_ - (AO_core_ ? naux_ * nbf_ * nbf_ : 0);
```
Does `rem_mem` need an alternate value depending on `release_AO_`?

yxie326 replied:
Yes, I think so. Changed it to `(AO_core_ && !release_core_AO_before_metric_ ? naux_ * nbf_ * nbf_ : 0)`.

loriab reviewed (commented):
@JonathonMisiewicz do you like `release_AO_` to `release_core_AO_before_metric_` before Yi applies the GH suggestions everywhere?

JonathonMisiewicz commented:
Proposed rename sounds good, thanks.

loriab commented on psi4/driver/procrouting/sapt/sapt_proc.py:
```diff
@@ -111,148 +114,162 @@ def run_sapt_dft(name, **kwargs):
+        else:
+            wfn_A = hf_wfn_A 
+            wfn_B = hf_wfn_B
+            data["DFT MONOMERA"] = hf_data["HF MONOMER A"] 
+            data["DFT MONOMERB"] = hf_data["HF MONOMER B"] 
```
In file https://github.com/psi4/psi4/blob/master/tests/sapt-compare/CMakeLists.txt, please add "quicktests" to the list like the below so that your changes go through the test suite.

`add_regression_test(sapt-compare "psi;sapt;cart;quicktests")`

loriab reviewed (commented):
Sorry about the pint errors; hopefully that gets resolved today. But at least some of the testing lanes are clean, so we can evaluate the changes.

loriab commented:
Apologies, please add `quick;` to the sapt-compare/test_input.py also.

loriab reviewed (approved):
lgtm, thanks for all the improvements!

JonathonMisiewicz reviewed (commented):
I'm trying to review this PR, but I'm puzzled by the logic. Looking at lines 119 and 147 of `sapt_proc.py`, it seems that it's possible to have `do_delta_hf = True` while `do_dft = False`. Is that right? The first command would imply to me that we _are_ doing DFT. Otherwise, delta HF is just zero.

What am I missing?

JonathonMisiewicz commented:
While I did not give this a very careful pass, I've given it enough of a pass to merge it in once the previous two comments I made are addressed.

@hokru, did you want another pass of this?

yxie326 commented on doc/sphinxman/source/sapt.rst:
```diff
@@ -533,9 +533,13 @@ estimated by scaling from the uncoupled value either by a fitted fixed
+scaling and using the uncoupled exchange-dispersion energy directly.
 
+.. warning:: Since Nov 2022, the defaults of options |sapt__sapt_dft_exch_disp_scale_scheme| and |sapt__sapt_dft_exch_disp_fixed_scale|
+             have been changed. Before, the former defaulted to ``FIXED`` with Hesselmann value of 0.686 for the latter. Now, the former defaults to ``DISP`` and should you instead select ``FIXED``, the default for the latter is the Xie value of 0.770. This might cause
```
Added the paper. Rephrased a bit on "Hesselmann appraoch" and "Xie approach"; I would call them "Podeszwa approach" and "Hesselmann approach" respectively instead (see references).

loriab commented on doc/sphinxman/source/sapt.rst:
```diff
@@ -530,12 +530,20 @@ term needs to be computed with orbital response for good accuracy,
-the former approach, ``DISP`` being the latter and ``NONE`` for not
-scaling and use the uncoupled exchange-dispersion energy directly.
-
+value initially suggested by [Hesselmann:2014:094107]_ and then revised
+by [Xie:2022:024801]_ for non-local Hartree-Fock (LHF) scheme. Alternatively,
+one can scale it by the ratio of
+coupled and uncoupled dispersion energy as suggested by [Podeszwa:2006:400]_. 
+This can be controlled by keyword |sapt__sapt_dft_exch_disp_scale_scheme|, 
+with ``FIXED`` being the Podeszwa approach as in (Psi's default prior 
+to Nov 2022), the Hesselmann approach (Psi's default after Nov 2022),
+and ``NONE`` for not scaling and using the uncoupled exchange-dispersion 
+energy directly.
```
```suggestion
value (suggested initially by [Hesselmann:2014:094107]_ for a local Hartree--Fock (LHF) formulation and then revised
by [Xie:2022:024801]_ for non-LHF) or
by the ratio of
coupled and uncoupled dispersion energy (suggested by [Podeszwa:2006:400]_ ). 
This can be controlled by keyword |sapt__sapt_dft_exch_disp_scale_scheme|, 
with ``FIXED`` using the Hesselmann/Xie approach (|PSIfours| default prior 
to Nov 2022), ``DISP`` using the Podeszwa approach (|PSIfours| default after Nov 2022),
or ``NONE`` for not scaling and using the uncoupled exchange-dispersion 
energy directly.
```

With all the cooks, this somehow lost its correctness (and sentencehood) :-)
